### ISSUE TITLE
Sporadic SIGSEGV (exit code -11) in Python when exiting the Alternator test

### ISSUE THREAD

**Issue by pwrobelse** [scylladb/scylladb#17564]
In CI run https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7054/testReport/ the following error occurred in `non-boost tests.alternator.run.release.2`:
```
=== TEST.PY SUMMARY START ===
Test exited with code -11
```

However, the log contains the following information:
```
============ 690 passed, 8 skipped, 53 xfailed, 1 xpassed in 48.07s ============
```
The log file can be found here: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7054/artifact/testlog/x86_64/release/alternator.run.2.log

**Comment by avikivity** [comment]
-11 is likely SIGSEGV. Probably it crashed during teardown.

**Comment by nyh** [comment]
We've seen this in the past, and @avikivity I don't think there's anything I can do about this issue, at least not now. As you can see, the Alternator tests **passes**, and printed "Alternator tests pass" - this is the very last thing that test/alternator/run does (printing `summary` is the very last thing that the atexit callback `cleanup_all()` does, see test/cql-pytest/run.py. If Python then died with SIGSEGV (signal 11) after priting this message, it's a Python bug. We've seen it happening in the past but never managed to reproduce it locally, or to get any information on how Python died or a coredump, or anything.

Please note that it's *not* Scylla which died. It's not even pytest which died. It's just the Python wrapper script test/alternator/run. As I said we've seen this issue in the past, fairly rarely, but unfortunately I can't find any issue about it. With the information I have now, it looks like a bug in Python that we cannot fix, especially not without any core dump from Python or something that can allow us to make a guess how we can perhaps work around this Python bug.

**Comment by nyh** [comment]
Here's an (Scylla-internal, unfortunately) thread about exactly the same SIGSEGV in Python from a year ago:
https://groups.google.com/a/scylladb.com/g/rnd-internal/c/XwLsbiOU5iM/m/mjAfxOKFAQAJ

We (@denesb and I) weren't able to explain it then, and I don't think we can make any progress now. We can leave this issue open in case we encounter it again the future, but don't think anyone should be assigned to it unless it becomes more frequent that one occurrence per year.

**Comment by mykaul** [comment]
We can (if we haven't done it already) ensure we run the latest stable Python version. I'm unsure if we do (https://github.com/scylladb/scylla-dtest/blob/d969bba3caeff377f66db0a0428c650f44f41d6f/jenkins_pipelines/docker_build.jenkinsfile#L25 ? https://github.com/scylladb/scylla-dtest/blob/d969bba3caeff377f66db0a0428c650f44f41d6f/Dockerfile#L1 ?)

**Comment by nyh** [comment]
> We can (if we haven't done it already) ensure we run the latest stable Python version. I'm unsure if we do (https://github.com/scylladb/scylla-dtest/blob/d969bba3caeff377f66db0a0428c650f44f41d6f/jenkins_pipelines/docker_build.jenkinsfile#L25 ? https://github.com/scylladb/scylla-dtest/blob/d969bba3caeff377f66db0a0428c650f44f41d6f/Dockerfile#L1 ?)

The problem is (which I mentioned to you last time we talked, but not very convincingly) is that the convenience of docker has a negative side-effect that it allows us to get "stuck" on ancient versions of development tools - in this case Python - and nobody knows what kind of mess will happen if we upgrade dtest to a newer Python because nobody runs dtests with a modern Python so there may be a lot of problems that nobody ever fixed. Because we waited so long, there can be a lot of mess. On the other hand, maybe it's not a lot of mess - somebody just needs to try.

**Comment by mykaul** [comment]
@fruch is doing a great job maintaining and updating the stack. I don't think we should always strive to get the latest-greatest (it has bugs too, just like we don't update dbuild every morning), but we shouldn't be stuck in the past as well.
I think updating to the latest patch release of a given Python release makes sense.

**Comment by fruch** [comment]
I'm not sure I understand why we are talking about the python version in dtest ?

Python is crashing inside dbuild, and that is using python 3.11 or even 3.12.

Sounds like it's not the first time, and also happened a year ago

I think @benipeled was working on a code to collect core dumps, maybe that would help more to get to the bottom of it.

**Comment by nyh** [comment]
I'm unassigning myself. As explained above, this is a crash in Python itself, which may be prompted by something used in the Alternator test suite but it's impossible to figure out why without a core dump - and even if we have one, I'm not sure we're in the business of debugging Python, and this bug appears to reproduce pretty rarely (I think we only saw it a handful of times). 
We can keep this issue open for future reference, but there's nothing I can do about it now.

**Comment by kostja** [comment]
This is most likely a duplicate of https://github.com/scylladb/scylladb/issues/18188 Anyhow when https://github.com/scylladb/scylladb/issues/18188 is implemented alternator/run won't be used.

**Comment by nyh** [comment]
I think we should close this issue. It's a Python bug of some sort, which hasn't happened in 6 months.

**Comment by temichus** [comment]
@kostja @nyh @pwrobelse 

Close this issue since it is not appened in 6 months, and tasks that @kostja mentioned are closed

**Comment by nyh** [comment]
Unfortunately, it still happens so I'm reopening this issue: happened today in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12990. test.py ran pytest on the test file alternator/test_streams.py, and when all the tests passed and pytest was shutting down, it crashed:
```
================ 36 passed, 11 xfailed, 1894 warnings in 19.96s ================
=== TEST.PY SUMMARY START ===
Test exited with code -11
```
This appears to be a very rare bug, and **not** Scylla's fault, but we should still be aware it exists because it causes CI failures :-(

**Comment by Annamikhlin** [comment]
seen also on 
https://jenkins.scylladb.com/job/scylla-6.2/job/next/87/testReport/junit/(root)/non-boost%20tests/Build___x86___Unit_Tests_x86___alternator_test_query_debug_2/

```
================= 27 passed, 2 skipped, 443 warnings in 10.51s =================
=== TEST.PY SUMMARY START ===
Test exited with code -11
```
full log: [alternator.test_query.2.log](https://github.com/user-attachments/files/17652932/alternator.test_query.2.log)

**Comment by nyh** [comment]
I didn't realize we have a second duplicate of this issue - https://github.com/scylladb/scylladb/issues/19973. We should probably close one of them, maybe I shouldn't have reopened this one :-(

**Comment by nyh** [comment]
Only a month after this issue was closed, a duplicate issue was created https://github.com/scylladb/scylladb/issues/19973 and started collecting more sightings of this issue. I'll close the duplicate issue, and let's keep this one open - it's not solved. I'm not saying we can fix it in any way - it appears to be some sort of bug in Python and/or boto3 - but we should at least know this problem exists and not create more duplicates of it.

**Comment by fruch** [comment]
@nyh why do you think it's boto3 ? just cause it happens on the alternator tests ?

if boto3 is suspect, I would look for code opening botocore session in multiple threads. boto3 opening session isn't thread safe, we run into some issue (not coredumps) in SCT when opening session in multiple threads at the same time.

again just a guess, without more details collected it's hard to pin it down...

**Comment by nyh** [comment]
I just ran `test/alternator/run test_query.py` (the last run to be reported above to crash on the test's exit) one thousand (!) times, and it didn't crash even once.
So whatever this crash is (my guess, but it's only a guess - a Python shutdown bug while some boto3 sockets or background threads or whatever are still alive) it is probably very rare and I still don't have a clue what causes it or how to avoid it.

**Comment by mykaul** [comment]
I assume (but unsure) we run the latest available Python on our CI instances?

**Comment by nyh** [comment]
@fruch I wrote you a long response which got deleted in the JumpCloud farce we had today, so I'll write a short response: 
1. I suspect boto3 or some library it uses (urllib? requests?) because we never saw these crashes on the cql-pytest tests which use the same Python and same pytest
2. It can't be related to threads because one of the test files that reproduced this crash, test_query.py (for example), does not use ANY threads whatsoever.
3. I suspect (without any proof) that boto3 has some threads or sockets open when shutting down, and somehow this can cause a crash during shutdown. This can't be a bug in any code written in Python (that can't cause a SIGSEGV), so it's a bug in some internal Python code related to sockets, threads, the GIL or something. I tried to Google for discussions of such a bug, couldn't find one.

**Comment by fruch** [comment]
> [@fruch](https://github.com/fruch) I wrote you a long response which got deleted in the JumpCloud farce we had today, so I'll write a short response:
> 
> 1. I suspect boto3 or some library it uses (urllib? requests?) because we never saw these crashes on the cql-pytest tests which use the same Python and same pytest
> 2. It can't be related to threads because one of the test files that reproduced this crash, test_query.py (for example), does not use ANY threads whatsoever.
> 3. I suspect (without any proof) that boto3 has some threads or sockets open when shutting down, and somehow this can cause a crash during shutdown. This can't be a bug in any code written in Python (that can't cause a SIGSEGV), so it's a bug in some internal Python code related to sockets, threads, the GIL or something. I tried to Google for discussions of such a bug, couldn't find one.

python code can cause segmentation fault indirectly, and if it's a fault in boto3, which version is being used ?

**Comment by fruch** [comment]
> I assume (but unsure) we run the latest available Python on our CI instances?

those tests are using the python version from within dbuild, i.e. python3.12

**Comment by nyh** [comment]
> python code can cause segmentation fault indirectly

I don't know what you mean by indirectly. My guess (and again, it's just a guess) is that there's a bug in some Python handling of background threads during shutdown or open sockets, or something, and the fact is that boto3 opens these threads or sockets or whatever - and cql doesn't (or opens fewer of them, or opens them differently, or something like that). Anyway, these are just guesses - I was never able to reproduce this crash locally, nor to get a core dump from someone else's crash.

>, and if it's a fault in boto3, which version is being used ?

I don't know, it's whatever boto3 is in dbuild. I see in install-dependencies.sh that we take the python3-boto3 package from Fedora. On my Fedora 40 (I don't know about dbuild), it is 1.35.54. I doubt the specific version of boto3 matters - we've been seeing this problem (very rarely) for years now.

**Comment by fruch** [comment]
sound kind of like: https://github.com/boto/boto3/issues/1062

but the reports there, the newer python version fixed the issue.
but still someone reported something close on python 3.11 (but seem to be Mac specific)

**Comment by nyh** [comment]
Seen again in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14456/testReport/junit/(root)/non-boost%20tests/alternator_test_update_expression_debug_2/ -
```
=== TEST.PY STARTING TEST alternator.test_update_expression.2 ===
...
================= 56 passed, 2 xfailed, 380 warnings in 11.48s =================
=== TEST.PY SUMMARY START ===
Test exited with code -11

=== TEST.PY SUMMARY END ===
```

As usual, all the test passed, pytest said that, and when the Python running pytest was existing, it had a SIGSEGV.

**Comment by fruch** [comment]
@nyh 

while looking as some crash in SCT, I was looking on how to get more specific information from such crashes

and seem like https://docs.python.org/3/using/cmdline.html#envvar-PYTHONFAULTHANDLER can help get out a python backtrace of where the signal is happening, that might help a bit.

**Comment by nyh** [comment]
I created a pull request https://github.com/scylladb/scylladb/pull/22341 to call boto3's client close() - which ends up calling urllib3's PoolManager's clear() - when tearing down the test fixures. If the bug is caused by connections left behind or something like this, **maybe** it will make this bug go away. But I can't be sure because I've never been able to reproduce this bug. But the fact we regularly see this crash in test/alternator and never ever saw it test/cqlpy makes me quite confident in my guess that this bug is specific to boto3 and in specifically on the things it leaves behind for Python to clean up during shutdown. If my pull request goes in and we'll have a year without reports of this issue, we'll know the fix worked. Let's hope that this will be the case.

**Comment by nyh** [comment]
> [@nyh](https://github.com/nyh)
> 
> while looking as some crash in SCT, I was looking on how to get more specific information from such crashes
> 
> and seem like https://docs.python.org/3/using/cmdline.html#envvar-PYTHONFAULTHANDLER can help get out a python backtrace of where the signal is happening, that might help a bit.

Interesting. If you think this helps (I don't know how to induce a Python crash to see it actually works ;-) - _EDIT: I see https://docs.python.org/3/library/faulthandler.html has an example way to induce it_), we could add this environment variable to Jenkins and it will be inherited by all the tests. That is, unless test.py clears the environment for the processes it calls (in the past, @kostja wanted to do such clearing because of specific environment variables that caused problems, I urged him not to do it, I hope it didn't get in).

**Comment by fruch** [comment]
> > [@nyh](https://github.com/nyh)
> > while looking as some crash in SCT, I was looking on how to get more specific information from such crashes
> > and seem like https://docs.python.org/3/using/cmdline.html#envvar-PYTHONFAULTHANDLER can help get out a python backtrace of where the signal is happening, that might help a bit.
> 
> Interesting. If you think this helps (I don't know how to induce a Python crash to see it actually works ;-) - _EDIT: I see https://docs.python.org/3/library/faulthandler.html has an example way to induce it_), we could add this environment variable to Jenkins and it will be inherited by all the tests. That is, unless test.py clears the environment for the processes it calls (in the past, [@kostja](https://github.com/kostja) wanted to do such clearing because of specific environment variables that caused problems, I urged him not to do it, I hope it didn't get in).

pytest doesn't do those thing out of the box (maybe some plugin can, it's really easy) but I wouldn't recommend it either, it can break lots of things if not handled correctly

in SCT we facing such thing, we introduce a check at the beginning to validate any environment variables related to the test are using in the way they are intended (it's no exhaustive as I would want, slowly covering known issues with missconfiguration)

**Comment by fruch** [comment]
@nyh 

I've trying to introduce it in SCT:
https://github.com/scylladb/scylla-cluster-tests/pull/9841

I'll keep you updated if it helped or not.

**Comment by fruch** [comment]
> Happened again in a new CI run https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15083/testReport/junit/(root)/non-boost%20tests/alternator_test_scan_debug_8/, so unfortunately it seems that https://github.com/scylladb/scylladb/commit/955ac1b7b794f71da528b5f4660af17a2ecd6900 (adding a close() of some boto3) didn't solve the problem. 
> 
> The new failure happened in the `test_scan` test file.
> 
> @fruch did you add any luck hunting the similar problems you said you had in SCT?

The change I did in SCT was to collect information about python crashes, I didn't yet run into a crash to see it in use.

Regardless if you suspect boto3, it might be something about threading, you shouldn't be creating sessions in multiple threads at the same time. boto3 isn't completely thread safe.

**Comment by nyh** [comment]
> > Happened again in a new CI run https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15083/testReport/junit/(root)/non-boost%20tests/alternator_test_scan_debug_8/, so unfortunately it seems that [955ac1b](https://github.com/scylladb/scylladb/commit/955ac1b7b794f71da528b5f4660af17a2ecd6900) (adding a close() of some boto3) didn't solve the problem.
> > The new failure happened in the `test_scan` test file.
> > [@fruch](https://github.com/fruch) did you add any luck hunting the similar problems you said you had in SCT?
> 
> The change I did in SCT was to collect information about python crashes, I didn't yet run into a crash to see it in use.
> 
> Regardless if you suspect boto3, it might be something about threading, you shouldn't be creating sessions in multiple threads at the same time. boto3 isn't completely thread safe.

There are no threads at all. For example, the latest failure https://github.com/scylladb/scylladb/issues/23338 was when running test/alternator/test_update_expression.py.  There's no threading, or asyncio or anything like that involved. Just a single boto3 session (the "dynamodb" fixture) for all tests.

**Comment by nyh** [comment]
@xtrey I just realized something: The failures in all the duplicates of this issue (like the most recent https://github.com/scylladb/scylladb/issues/23338) are not in any specific test function, but when shutting down the test for the whole file. But in a recent patch, you eliminated this extra log file for the entire test file, because it was considered redundant.

So it seems to me that even if this Python shutdown bug still exists, and a "pytest" run fails once in a while, we will no longer know this because we'll not have a separate output file for this failure. Am I correct?

**Comment by xtrey** [comment]
@nyh Reports will be green, but CI will be red. `test.py` still will fail the test because the exit code is not in a valid exit codes. So we will need to check the console log to get the information that it failed because of this issue. This is some issue with python, I suppose. pytest finishes all that we need from him correctly, we have generated JUnit report.

> There's no threading, or asyncio or anything like that involved. Just a single boto3 session (the "dynamodb" fixture) for all tests.

This is not correct. There are 6 simultaneous tests are executed. In the suite.yaml there is a parameter pool_size that is responsible for that.

**Comment by nyh** [comment]
> > There's no threading, or asyncio or anything like that involved. Just a single boto3 session (the "dynamodb" fixture) for all tests.
> 
> This is not correct. There are 6 simultaneous tests are executed. In the suite.yaml there is a parameter pool_size that is responsible for that.

It is correct - 

First of all, this issue started when Alternator tests were run from a test/alternator/run script, and the problem already existed.

But even today, before your "everything will run from pytest" makeover, test.py still runs each pytest as a separate **process** (test.py runs a separate pytest for each individual test file, IIRC) . Inside this pytest process, there are no threads (we do have a couple of tests that use threads, but the failures we saw were not in those files - the files that failed did not have any usage of threads at all).

The fact that test.py runs 6 "pytest" processes in parallel should not matter at all - these processes do not share anything.

**Comment by ScyllaPiotr** [comment]
> Reports will be green, but CI will be red. `test.py` still will fail the test because the exit code is not in a valid exit codes.

Looking at the number of messages, this cost us much effort already.

Could we filter out the CI failures that the log output of contains a valid test summary line? I mean a line that matches a pattern similar to this one: `================= 56 passed, 2 xfailed, 380 warnings in 11.48s =================`

**Comment by ScyllaPiotr** [comment]
> [@nyh](https://github.com/nyh) enabled traceback specifically to such cases to understand what is going wrong. 

Yes I know, but I'm asking if we can **ignore** (instead of understanding what is going wrong) __the error code of test part of the CI__ whenever the log output contains the aforementioned line.

> [@ScyllaPiotr](https://github.com/ScyllaPiotr) Please share the link to the job where it failed for further investigation.

There was no new job, I copied it from [here](https://github.com/scylladb/scylladb/issues/17564#issuecomment-2594710519).

**Comment by xtrey** [comment]
Yes, we can ignore. But I'm trying to understand how it can cost much effort if there were no new issues? I'm just trying to understand how much effort we need to implement this issue against how much effort we are wasting looking at the log files. And for now, it looks like we are wasting zero, since there were no new manifestation of this bug.

**Comment by nyh** [comment]
I agree with both @xtrey and @ScyllaPiotr:

1. As I mentioned in another thread, this issue happens very rarely, last time we saw it was two months ago, and since then I already add more printouts which perhaps will help us understand the problem when it next happens, and until then I don't think we should do anything.

2. But, when this issue happens again and **if** the new printouts won't help us, **then** we can consider @ScyllaPiotr's idea - indeed we can change test.py to check specifically if a "pytest" process it ran died with a signal 11, and if it did, and if the log file ends with signs that all the tests ran before the crash, then test.py can simply ignore this error (you can still have erros in the individual tests, of course). This would be a valid fix - we _really_ don't care if pytest successfully shut down or didn't after it reported all the tests to have run.

**Comment by ScyllaPiotr** [comment]
Agreed.

Just to be more clear,

![Image](https://github.com/user-attachments/assets/3d0127fc-d4fe-4b0f-a49e-f01e48f6f48f)

That's the effort I have in mind.

**Comment by nyh** [comment]
Unfortunately, this error happened again, in a CI run for the 2025.1 branch.
Since https://github.com/scylladb/scylladb/pull/23307 is not in yet (I was surprised to learn this now), we don't have the debugging information I hoped we will have when this failed again :-(

**Comment by nyh** [comment]
Seen again, this time in a Scylla-enterprise run https://jenkins.scylladb.com/job/scylla-enterprise/job/next/3465/artifact/testlog/x86_64/release/alternator.test_table.1.log

Unfortunately, I can't find there any output from https://github.com/scylladb/scylladb/pull/23340 (which is in the enterprise-next already). I don't know what to make of this - perhaps the Python debugging didn't work, or for some reason not captured into the log. So I'm still completely in the dark why this is happening.

Perhaps it's about time to consider @ScyllaPiotr's idea - change the test.py or Jenkins to just ignore this specific error (pytest getting signal 11 **after** it already reported that all tests it ran were successful). The current situation once in a while wastes time of one of us (@lsfreitas was the latest "victim") who needs to figure out how come the full test run failed, but Jenkins says all the individual tests were actually successful.

**Comment by fruch** [comment]
BTW I had success in SCT with capturing the coredumps and analyzing it.

in my case it was a mixture of a callback coming from native c code, during python shutdown (it was part of python-driver), and code on SCT that cleared the atexit callbacks to prevent from some buggy code to stuck the shutdown.

I would guess it's python shutdown as well in your case

in SCT we have this code:
https://github.com/scylladb/scylla-cluster-tests/commit/1f87013f63c7d1a1cd71d2e53d06e185ffdaedc9

that helps to figure out what is still running when the tests are finishing, it might help to flush out all kind of rough processes or threads, that wasn't cleared out.

**Comment by m-szymon** [comment]
It happened also here: https://github.com/scylladb/scylladb/issues/24322.

**Comment by ScyllaPiotr** [comment]
Another one: #24847.
Can we update Python used by the CI? I saw that we recently moved the build toolchain to F42, do we use default Python version for Fedora (and it just got an update because of the move) or a custom one?

**Comment by ScyllaPiotr** [comment]
Please remove the word _sporadic_ from the issue name 😉 

Jokes aside,
> @nyh enabled traceback specifically to such cases to understand what is going wrong.

and

> But in a recent patch https://github.com/scylladb/scylladb/pull/23237, you eliminated this extra log file for the entire test file, because it was considered redundant.
> So it seems to me that even if this Python shutdown bug still exists, and a "pytest" run fails once in a while, we will no longer know this because we'll not have a separate output file for this failure. Am I correct? Or, maybe we will get a Jenkins failure without any specific failed test - and if we look at the Jenkins "console log" we'll see the failed pytest's output?

I don't see any extra logs on the [console of the latest run](https://jenkins.scylladb.com/job/scylla-2025.2/job/next/107/consoleFull).
Plus

> I had success in SCT with capturing the coredumps and analyzing it.

So, @fruch are you aware of a way to access the machines used for [the latest build](https://jenkins.scylladb.com/job/scylla-2025.2/job/next/107/) to see if the logs/dumps that @nyh enhanced are there?

Alternatively, can we could undo the [patch](https://github.com/scylladb/scylladb/pull/23237) that is suspected to purge the file we're looking for, but [plain `git revert` is probably not an option](https://github.com/scylladb/scylladb/issues/23220#issuecomment-2713445762).

Or, we can just ignore [these errors](https://github.com/scylladb/scylladb/issues/17564#issuecomment-2796301025).

**Comment by fruch** [comment]
> Please remove the word _sporadic_ from the issue name 😉
> 
> Jokes aside,
> 
> > [@nyh](https://github.com/nyh) enabled traceback specifically to such cases to understand what is going wrong.
> 
> and
> 
> > But in a recent patch [#23237](https://github.com/scylladb/scylladb/pull/23237), you eliminated this extra log file for the entire test file, because it was considered redundant.
> > So it seems to me that even if this Python shutdown bug still exists, and a "pytest" run fails once in a while, we will no longer know this because we'll not have a separate output file for this failure. Am I correct? Or, maybe we will get a Jenkins failure without any specific failed test - and if we look at the Jenkins "console log" we'll see the failed pytest's output?
> 
> I don't see any extra logs on the [console of the latest run](https://jenkins.scylladb.com/job/scylla-2025.2/job/next/107/consoleFull). Plus
> 
> > I had success in SCT with capturing the coredumps and analyzing it.
> 
> So, [@fruch](https://github.com/fruch) are you aware of a way to access the machines used for [the latest build](https://jenkins.scylladb.com/job/scylla-2025.2/job/next/107/) to see if the logs/dumps that [@nyh](https://github.com/nyh) enhanced are there?
> 
> Alternatively, can we could undo the [patch](https://github.com/scylladb/scylladb/pull/23237) that is suspected to purge the file we're looking for, but [plain `git revert` is probably not an option](https://github.com/scylladb/scylladb/issues/23220#issuecomment-2713445762).
> 
> Or, we can just ignore [these errors](https://github.com/scylladb/scylladb/issues/17564#issuecomment-2796301025).

talk with @yaronkaikov, about accessing those.
in SCT we added a whole machinary to collect the dumps from each runs, since the builder are disposable

**Comment by m-szymon** [comment]
Another occurence: https://github.com/scylladb/scylladb/pull/25381#issuecomment-3166128916

**Comment by m-szymon** [comment]
Same error: https://github.com/scylladb/scylladb/issues/25531

**Comment by ScyllaPiotr** [comment]
@yaronkaikov Am I correct to observe that the issue has not occurred since we started using Fedora 42? Maybe the new version of Python stopped crashing.

**Comment by yaronkaikov** [comment]
> [@yaronkaikov](https://github.com/yaronkaikov) Am I correct to observe that the issue has not occurred since we started using Fedora 42? Maybe the new version of Python stopped crashing.

I don't know, but the timelines make sense. we moved to Fedora 42 (in cloud a well ) on the 21 of August

**Comment by fruch** [comment]
> [@yaronkaikov](https://github.com/yaronkaikov) Am I correct to observe that the issue has not occurred since we started using Fedora 42? Maybe the new version of Python stopped crashing.

it's never the kernel fault :)

it's probably one of the packages fault, and in SCT I have evidence of scylla-driver that can cause such issues
https://github.com/scylladb/python-driver/issues/524

we are using newer python version in SCT and still hitting it from time to time.

**Comment by nyh** [comment]
We've been seeing issue for several **years**, and it's only been a month since the last report. Let's not break out the champagne just yet...

**Comment by m-szymon** [comment]
Is backport 2025.1 using Fedora 42?
https://jenkins.scylladb.com/job/scylla-2025.1/job/scylla-ci/657/testReport/junit/(root)/non-boost%20tests/alternator_test_scan_debug_9/
Anyway @ScyllaPiotr you broke my build by thinking about this issue.

**Comment by temichus** [comment]
Remove my team assignment as this issue is in the Alternetor team

**Comment by radoslawcybulski** [comment]
@fruch which version of python are we using? Can we bump up to 3.14? There're changes (in 3.14) around `PyEval_RestoreThread`, which might resolve the issue (namely instead of deletion the thread is blocked forever).

**Comment by fruch** [comment]
> @fruch which version of python are we using? Can we bump up to 3.14? There're changes (in 3.14) around `PyEval_RestoreThread`, which might resolve the issue (namely instead of deletion the thread is blocked forever).

@avikivity is working on updating the toolchain, and it would have 3.14 in it.

Even though I doubt it has a fix for this, it's not mentioned in the original bug

**Comment by bhalevy** [comment]
Seen again in https://jenkins.scylladb.com/job/scylla-2025.1/job/scylla-ci/770/testReport/junit/(root)/non-boost%20tests/Tests___Unit_Tests___alternator_test_scan_debug_3/ (#28067)

**Comment by nyh** [comment]
@bhalevy note that the new failure is in an old branch (2025.1), so if we had above a theory that maybe some Python upgrades or something made the problem go away then this new failure doesn't disprove that theory.

**Comment by bhalevy** [comment]
> [@bhalevy](https://github.com/bhalevy) note that the new failure is in an old branch (2025.1), so if we had above a theory that maybe some Python upgrades or something made the problem go away then this new failure doesn't disprove that theory.

Agreed

**Comment by ScyllaPiotr** [comment]
Demoting priority until the problem reappears.
